### PR TITLE
fix #1648: support submit storm jar via rest interface

### LINKED ISSUE #1648: support submitting storm jar on Web UI

### PR DESCRIPTION
**Pull request by stanleyxu2005**

### CODE REVIEW

File: integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala [hunk 1]
```diff
@@ -75,7 +75,11 @@ class RestClient(host: String, port: Int) {
     listApps().filter(_.status == MasterToAppMaster.AppMasterActive)
   }
 
-  def submitApp(jar: String, args: String = "", config: String = ""): Int = try {
+  def getNextAvailableAppId(): Int = {
+    listApps().length + 1
```
Comment (huafengw) [inline integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: What if user has killed some apps?

Comment (huafengw) [inline integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: I think it should be listApps().map(_.appId).max + 1

**Comment by manuzhang** [comment]
submit storm application through UI failed with the following error. The same example could be run successfully through command line.

```
[ERROR] [01/22/2016 14:30:21.968] [OneForOneStrategy] cannot assign instance of scala.collection.immutable.List$SerializationProxy to field io.gearpump.util.Graph.edgeList of type scala.collection.immutable.List in instance of io.gearpump.util.Graph
akka.actor.ActorInitializationException: exception during creation
        at akka.actor.ActorInitializationException$.apply(Actor.scala:166)
        at akka.actor.ActorCell.create(ActorCell.scala:596)
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:456)
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478)
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:263)
        at akka.dispatch.Mailbox.run(Mailbox.scala:219)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.ClassCastException: cannot assign instance of scala.collection.immutable.List$SerializationProxy to field io.gearpump.util.Graph.edgeList of type scala.collection.immutable.List in instance of io.gearpump.util.Graph
        at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2127)
        at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:1305)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2003)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1921)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1798)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1350)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
        at akka.serialization.JavaSerializer$$anonfun$1.apply(Serializer.scala:136)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
        at akka.serialization.JavaSerializer.fromBinary(Serializer.scala:136)
        at akka.serialization.Serializer$class.fromBinary(Serializer.scala:60)
        at akka.serialization.JavaSerializer.fromBinary(Serializer.scala:120)
        at io.gearpump.cluster.UserConfig.getValue(UserConfig.scala:116)
        at io.gearpump.streaming.appmaster.AppMaster.io$gearpump$streaming$appmaster$AppMaster$$getUpdatedDAG(AppMaster.scala:310)
        at io.gearpump.streaming.appmaster.AppMaster$$anonfun$3.apply(AppMaster.scala:78)
        at io.gearpump.streaming.appmaster.AppMaster$$anonfun$3.apply(AppMaster.scala:77)
        at akka.actor.TypedCreatorFunctionConsumer.produce(Props.scala:343)
        at akka.actor.Props.newActor(Props.scala:252)
        at akka.actor.ActorCell.newActor(ActorCell.scala:552)
        at akka.actor.ActorCell.create(ActorCell.scala:578)
        ... 9 more
```

File: project/Build.scala [hunk 2]
```diff
@@ -270,7 +270,11 @@ object Build extends sbt.Build {
   lazy val services: Project = services_full.jvm.
     settings(serviceJvmSettings: _*).
     settings(compile in Compile <<= (compile in Compile) dependsOn (fastOptJS in (serviceJS, Compile))).
-    dependsOn(streaming % "test->test;compile->compile", daemon % "test->test;compile->compile;provided")
+    dependsOn(
+      streaming % "test->test;compile->compile",
+      daemon % "test->test;compile->compile;provided",
+      storm % "test->test;compile->compile;provided"
```
Comment (clockfly) [inline project/Build.scala]: This seems creating too many dependencies.

Comment (clockfly) [inline project/Build.scala]: And storm jar and daemon jar will conflict with each other.

File: services/dashboard/views/apps/apps.js [hunk 3]
```diff
@@ -34,16 +34,25 @@ angular.module('dashboard')
...
+        submitWindow.$promise.then(submitWindow.show);
+      };
+
+      $scope.openSubmitStormAppDialog = function() {
+        submitWindow.$scope.isStormApp = true;
         submitWindow.$promise.then(submitWindow.show);
       };
 
       $scope.composeMenuOptions = [{
+        text: '<i class="glyphicon glyphicon-none"></i> <b>Submit User Application</b>',
```
Comment (clockfly) [inline services/dashboard/views/apps/apps.js]: Maybe Submit Gearpump Jar Application
Submit Storm Jar better？

Comment (clockfly) [inline services/dashboard/views/apps/apps.js]: Where is compose DAG? Also in the dropdown list?

File: integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala (hunk 1 above)
Comment (stanleyxu2005) [inline integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: Your version looks better. Actually `listApps` will return all apps including killed.

File: services/dashboard/views/apps/apps.js (hunk 3 above)
Comment (stanleyxu2005) [inline services/dashboard/views/apps/apps.js]: There will be three options. Compose DAG is there. 

I think "Submit *Foo* Application" will be better. JAR is actually a part of an application. 
Regarding "Submit Gearpump Application" vs "Submit User Application", I think Gearpump is better than user, as we might submit a Gearpump built-in example manually, which is not an user self developed application.

**Comment by manuzhang** [comment]
@stanleyxu2005 

There is no need for `services` to depend on `storm`.
We can add storm classpath while launching the "StormRunner" process.

* load `StormRunner` with the full class name instead of `StormRunner.getClass`
* add the storm classpath `System.getProperty("gearpump.home") + "/lib/storm/*"`

File: services/jvm/src/main/scala/io/gearpump/services/MasterService.scala [hunk 4]
```diff
@@ -184,49 +198,97 @@ object MasterService {
...
-        options :+= s"-D${Constants.GEARPUMP_CUSTOM_CONFIG_FILE}=${userConf.get.getPath}"
+      val jar = fileMap.get("jar")
+      if (jar.isEmpty) {
+        throw new IOException("JAR file not supplied")
       }
 
-      val arguments = Array("-jar", jar.getPath) ++ extraArgs
-      val mainClass = AppSubmitter.getClass.getName.dropRight(1)
-      val process = Util.startProcess(options, Util.getCurrentClassPath, mainClass, arguments)
+      val stormLibs = Array(System.getProperty("gearpump.home") + "/lib/storm/*")
```
Comment (manuzhang) [inline services/jvm/src/main/scala/io/gearpump/services/MasterService.scala]: 1. I think only storm application requires this classpath. 
2. we have a `util.Constants.GEARPUMP_HOME`

File: services/jvm/src/main/scala/io/gearpump/services/MasterService.scala [hunk 5]
```diff
@@ -184,49 +198,97 @@ object MasterService {
...
+      val process = Util.startProcess(
+        options = clusterOptions(sysConfig, fileMap.get("gear")),
+        classPath = Util.getCurrentClassPath ++ stormLibs,
```
Comment (manuzhang) [inline services/jvm/src/main/scala/io/gearpump/services/MasterService.scala]: Services' classpath should not be used for application

**Comment by manuzhang** [comment]
@stanleyxu2005 you can refer to generated `output/target/pack/bin/storm` for an example of launching `StormRunner`.

File: integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala (hunk 1 above)
Comment (stanleyxu2005) [inline integrationtest/core/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: Your version looks better. as long as `listApps` does not return an empty Array. Actually `listApps` will return all apps including killed. `listRunningApps` will only return running apps, which we cannot use.

File: core/src/main/scala/io/gearpump/util/Util.scala [hunk 6]
```diff
@@ -45,6 +45,24 @@ object Util {
...
+      homeDir + "conf",
+      libHomeDir + "daemon/*",
+      libHomeDir + "yarn/*",
+      libHomeDir + "*"
+    )
+  }
+
+  def getStormApplicationClassPath: Array[String] = {
+    getUserApplicationClassPath ++ Array(
+      libHomeDir + "storm/*"
```
Comment (clockfly) [inline core/src/main/scala/io/gearpump/util/Util.scala]: Can you avoid putting any storm or yarn stuff in project core?

File: services/jvm/src/main/scala/io/gearpump/services/util/HttpUtil.scala [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package io.gearpump.services.util
+
+object HttpUtil {
+
+  /**
+   * Return a space separated arguments as an array.
+   */
+  def spaceSeparatedArgumentsToArray(str: String): Array[String] = {
```
Comment (clockfly) [inline services/jvm/src/main/scala/io/gearpump/services/util/HttpUtil.scala]: Why this should be a util? Anyone else using it?

File: core/src/main/scala/io/gearpump/util/Util.scala (hunk 6 above)
Comment (stanleyxu2005) [inline core/src/main/scala/io/gearpump/util/Util.scala]: Updated. Moved all you pointed code back to `MasterService.scala`

**Comment by manuzhang** [comment]
compile failed 
```
[error] /home/travis/build/gearpump/gearpump/services/jvm/src/main/scala/io/gearpump/services/MasterService.scala:223: not enough arguments for method submitAndDeleteTempFiles: (mainClass: String, argsArray: Array[String], fileMap: Map[String,java.io.File], classPath: Array[String], systemConfig: com.typesafe.config.Config, userConfigFile: Option[java.io.File])Boolean.

[error] Unspecified value parameter systemConfig.

[error]     submitAndDeleteTempFiles(

[error]  
```

File: services/jvm/src/main/scala/io/gearpump/services/MasterService.scala [hunk 8]
```diff
@@ -184,49 +198,124 @@ object MasterService {
...
+      }
     }
   }
 
+  /**
+   * Return Java options for gearpump cluster
+   */
+  private def clusterOptions(systemConfig: Config, userConfigFile: Option[File]): Array[String] = {
+    var options = Array(
+      s"-D${Constants.GEARPUMP_HOME}=${systemConfig.getString(Constants.GEARPUMP_HOME)}",
```
Comment (manuzhang) [inline services/jvm/src/main/scala/io/gearpump/services/MasterService.scala]: do we need `"-Djava.rmi.server.hostname=localhost"` ? I see they are set for `StormRunner` in pack

Comment (stanleyxu2005) [inline services/jvm/src/main/scala/io/gearpump/services/MasterService.scala]: If it is really necessary, can you later move it to `StormRunner`? I really don't like a strong hypothese to include such a java option here.

**Comment by manuzhang** [comment]
+1

**Pull request merged**